File: problems/sort-array-by-parity-ii.cpp
class Solution {
public:
    vector<int> sortArrayByParityII(vector<int>& nums) {
        int odd = 0;
        int n = nums.size();
        // First, transfer all odd numbers in the left side.
        for(int i = 0; i < n; i++){
            if(odd >= n) break;
            if(nums[i] % 2){
                swap(nums[i], nums[odd++]);
            }
        }
        // Now, just swap between sides every other element
        int i = 0; 
        int j = n - 1;
        while(i < j){
            swap(nums[i], nums[j]);
            i += 2;
            j -= 2;
        }
        return nums;
    }
};

/*
    Complexity analysis:
    Time: O(n), where `n` is the length of the given array, as we need to scan every element at least once.
    Space: O(1), since we do not use any extra space. 
*/

File: problems/sort-array-by-parity-ii.java
class Solution {
    public int[] sortArrayByParityII(int[] arr) {
       int even = 0;
        for (int i = 1; i < arr.length; i += 2) {
            if (arr[i] % 2 == 0) {
                while (arr[even] % 2 == 0)
                    even += 2;

                int temp = arr[even];
                arr[even] = arr[i];
                arr[i] = temp;
            }

        }
        return arr;
    }
}

/*
    Complexity analysis:
    Time: O(n), where `n` is the length of the given array, as we need to scan every element at least once.
    Space: O(1), since we do not use any extra space. 
*/
